### PR TITLE
Umcs 453/upl extend customer data

### PR DESCRIPTION
**Pull request by Ryouzanpaku**
- Add companyInfo.owner property.
- Add companyInfo.companyType property.
- Add shippingType to `Address` usable for any kind of address.

### CODE REVIEW

File: src/Resources/Customer.php [hunk 1]
```diff
@@ -343,6 +343,7 @@ public function handleResponse(stdClass $response, $method = HttpAdapterInterfac
     {
         if (isset($response->companyInfo) && $this->companyInfo === null) {
             $this->companyInfo = new CompanyInfo();
+            $this->companyInfo->handleResponse($response->companyInfo);
```
Comment (sixer1182) [inline src/Resources/Customer.php:346]: The companyInfo was synched before, why is this necessary?

File: src/Resources/EmbeddedResources/CompanyInfo.php [hunk 2]
```diff
@@ -121,6 +129,58 @@ public function setCommercialSector(string $commercialSector): CompanyInfo
...
+
+    //<editor-fold desc="Overridable methods">
+
+    /**
+     * {@inheritDoc}
+     */
+    public function handleResponse(stdClass $response, $method = HttpAdapterInterface::REQUEST_GET): void
+    {
+        if (isset($response->owner) && $this->owner === null) {
+            $this->owner = new CompanyOwner();
```
Comment (sixer1182) [inline src/Resources/EmbeddedResources/CompanyInfo.php:180]: Does it sync properly if the customer is fetched?

File: src/Resources/EmbeddedResources/CompanyOwner.php [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    public function setLastname(?string $lastname): CompanyOwner
+    {
+        $this->lastname = $lastname;
+        return $this;
+    }
+
+    /**
+     * @return string|null
+     */
+    public function getBirthdate(): ?string
```
Comment (sixer1182) [inline src/Resources/EmbeddedResources/CompanyOwner.php:83]: DateTime
```suggestion
    public function getBirthdate(): ?DateTime
```

File: src/Resources/EmbeddedResources/CompanyOwner.php [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    {
+        return $this->birthdate;
+    }
+
+    /**
+     * @param string|null $birthdate
+     *
+     * @return CompanyOwner
+     */
+    public function setBirthdate(?string $birthdate): CompanyOwner
```
Comment (sixer1182) [inline src/Resources/EmbeddedResources/CompanyOwner.php:93]:
```suggestion
    public function setBirthdate(?DateTime $birthdate): CompanyOwner
```

File: test/Fixtures/CustomerFixtureTrait.php [hunk 5]
```diff
@@ -122,12 +124,20 @@ public function getMinimalRegisteredB2bCustomer(): Customer
      */
     public function getMaximalRegisteredB2bCustomer(): Customer
     {
-        return $this->getMinimalRegisteredB2bCustomer()
+        $owner = (new CompanyOwner())
+            ->setFirstname('firstname')
+            ->setLastname('lastname')
+            ->setBirthdate('1999-01-01');
```
Comment (sixer1182) [inline test/Fixtures/CustomerFixtureTrait.php:130]: DateTime as defined in Tech Spec.
Is there a reason as to why not using DateTime?

File: test/Fixtures/CustomerFixtureTrait.php [hunk 6]
```diff
@@ -159,6 +169,7 @@ public function getShippingAddress(): Address
             ->setZip('60386')
             ->setCity('Frankfurt am Main')
             ->setCountry('DE')
-            ->setState('DE-BO');
+            ->setState('DE-BO')
+            ->setShippingType(ShippingTypes::EQUALS_BILLING);
```
Comment (sixer1182) [inline test/Fixtures/CustomerFixtureTrait.php:173]: Old tests should not be changed.
This parameter is present for upl invoice only

File: test/Fixtures/CustomerFixtureTrait.php [hunk 7]
```diff
@@ -122,12 +124,20 @@ public function getMinimalRegisteredB2bCustomer(): Customer
...
+            ->setBirthdate('1999-01-01');
+
+        $customer = $this->getMinimalRegisteredB2bCustomer()
             ->setShippingAddress($this->getShippingAddress())
             ->setSalutation(Salutations::MR)
             ->setMobile('+49172123456')
             ->setPhone('+4962216471100')
             ->setBillingAddress($this->getBillingAddress());
+
+        $customer->getCompanyInfo()->setOwner($owner);
```
Comment (sixer1182) [inline test/Fixtures/CustomerFixtureTrait.php:139]: Old tests should not be changed.
This parameter is present for upl invoice only.

File: test/integration/CustomerTest.php [hunk 8]
```diff
@@ -89,6 +89,21 @@ public function maxCustomerCanBeCreatedAndFetched(): Customer
         return $customer;
     }
 
+    /**
+     * Verify shipping type can be set for shipping address of customer resource.
+     *
+     * @test
+     */
+    public function customerWithShippingTypeCanBeCreateedAndFetched()
```
Comment (sixer1182) [inline test/integration/CustomerTest.php:97]:
```suggestion
    public function customerWithShippingTypeCanBeCreatedAndFetched()
```

**Review by sixer1182 — CHANGES_REQUESTED** [review]
There are some deviations from the tech spec.
Also no integration tests for company owner.
As well as other questions.

File: test/Fixtures/CustomerFixtureTrait.php (hunk 7 above)
Comment (Ryouzanpaku) [reply test/Fixtures/CustomerFixtureTrait.php:139]: This is about updating the maximum customer fixture which i think should be updated. This is independent of a specific payment method.
This fixture is used for customer tests only so far.

Comment (sixer1182) [reply test/Fixtures/CustomerFixtureTrait.php:139]: but you do not make sure "old" implementations still work.

File: test/Fixtures/CustomerFixtureTrait.php (hunk 6 above)
Comment (sixer1182) [reply test/Fixtures/CustomerFixtureTrait.php:173]: but you wont see if its absence would cause an error

Comment (sixer1182) [reply test/Fixtures/CustomerFixtureTrait.php:173]: there are two types of maximum customers now, since upl is a case that does not seemlesly fit into the rest of the papi

**Review by sixer1182 — APPROVED** [review]
